src/main/scala/com/fulcrumgenomics/util/miseq/SampleSheet.scala

```diff
@@ -0,0 +1,152 @@
+  val Lane: String           = "Lane"
+  val R1BarcodeBases: String = "R1_Barcode_Bases"
+  val R2BarcodeBases: String = "R2_Barcode_Bases"
+  val I7Bases: String        = "Index"
+  val I5Bases: String        = "Index2"
+
+  /** Attempts to clean special characters from a sample id or name as Illumina's software does. */
+  def cleanMiseqSampleId(id: String): String = {
+    id.map {
+      case '#' => ' '
+      case '_' | '+' | ' ' => '-'
+      case c => c
+    }
+  }
+
+  /** Gets the trimmed value of the given key (`name`) from the map.  If empty or not found, returns None */
+  private def getStringField(sampleDatum: Map[String, String], name: String): Option[String] = {
+    sampleDatum.get(name.toUpperCase) match {
+      case None => None
+      case Some(str) if str.isEmpty => None
+      case Some(str) => Some(str.trim)
+    }
+  }
+
+  private def getIntField(sampleDatum: Map[String, String], name: String): Option[Int] = {
+    getStringField(sampleDatum, name).map(_.toInt)
+  }
+
+  private[miseq] val SplitRegex = ",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)"
+}
+
+/** Create a sample sheet from the given file (typically 'SampleSheet.csv'). If the library id is omitted, the
+  * sample name will be used in its stead. */
+class SampleSheet(sampleSheet: Path, lane: Option[Int] = None) extends Iterable[Sample] {
+  import SampleSheet.SplitRegex
+  private val samples: ListBuffer[Sample] = new ListBuffer[Sample]()
+
+  getSampleData(sampleSheet).zipWithIndex.foreach {
+    case (sampleDatum, sampleOrdinal) => this.samples.append(makeSample(sampleOrdinal+1, sampleDatum))
+  }
+
+  def iterator: Iterator[Sample] = this.samples.iterator
+
+  override def size: Int = this.samples.size
+
+  def get(index: Int): Sample = this.samples(index)
+
+  /** Reads in the the Sample data only, one per row */
+  private def getSampleData(sampleSheet: Path): List[Map[String, String]] = {
+    // ignore data pre-"[Data]"
+    val (preData, postData) = Source.fromFile(sampleSheet.toFile).getLines().toList.span(!_.startsWith("[Data]"))
+
+    // get the header (keys) (skip "[Data]")
+    val header = (postData.drop(1).headOption match {
+      case Some(line) => line.split(SplitRegex, -1) // NB: include trailing empty strings
+      case None => throw new IllegalArgumentException("Could not find the header for sample data.")
+    }).map(_.toUpperCase.trim)
+
+    // get the rows (values), so skip "[Data]" and the header
+    val lineNumber = preData.size + 2 // 0-based
+    val sampleDatums = postData.drop(2).zipWithIndex.map { case (line, rowNumber) =>
+      val values = line.split(SplitRegex, -1)
+      // check we have the correct # of columns
+      if (values.size != header.length) {
+        throw new IllegalArgumentException(s"# of columns in the header and current row do not match ('${header.length}' != '${values.size}'): row #${rowNumber + 1} line number #${lineNumber + rowNumber + 1}")
```

Comment:
Would probably be easier on most folks to just spit out the offending line in the error rather than the line number.  Not to mention it would make your life simpler here.